### ISSUE TITLE
Consistency with object literals

### ISSUE THREAD

**Issue by ljharb** [alex-weej/es-extensible-collection-literal#1]
I think the syntax is very confusing if the LHS of a key/value pairing is an identifier - I'd suggest tweaking it so that it matches object literal syntax: it's a string, unless it's inside brackets, where it's computed just like in object literals and `class`.

**Comment by ljharb** [comment]
(I've read your FAQ about it; but I don't think "i have to type lots of brackets" is sufficient to overcome "it's different than similar-looking constructs everywhere else in the language")

**Comment by alexweej** [comment]
I disagree, and I went to quite some lengths to explain the manifold reasons why.

Not sure how to proceed with this as an opinion...!

**Comment by ljharb** [comment]
If this ever reached the committee, it'd be useful to get a sense of what the entire group thinks; but in past discussions we have very much prioritized consistency with existing syntax and idioms.

**Comment by bathos** [comment]
This surprised me too, though I haven’t spent enough time thinking about it to assign much weight to what’s really just my instinct to want to avoid new asymmetries. However I do wonder if it would violate the colon clause in "Forbidden Extensions":

> The Syntactic Grammar must not be extended in any manner that allows the token : to immediately follow source text that matches the BindingIdentifier nonterminal symbol.

If the usual object literal production is used here, I don’t think it violates that. If it is a new production with arbitrary expressions prior to the colon, I think it might, but I’m not sure.

**Comment by alexweej** [comment]
If it comes to it, a compromise of *requiring* `[]` could be reasonable.

```js
Immutable.Map#{
    ["one"]: 1,
    ["two"]: 2,
    ...
}

Immutable.Map#{
    [1]: "one",
    [2]: "two",
    ...
}
```

 It has the benefits that:

 * It aligns with object literals (a subset of the syntax). Consistency: check
 * String keys are always written like regular JS, not magical barewords with special exceptions. No special treatment is given to string keys.

Syntax highlighters would hopefully do a decent job of differentiating the `[]` surrounding a key from the `[]` delineating a sequence literal or Array literal, mitigating somewhat the concern about the ambiguity of:

```js
MyMap#{
    [[1, 2, 3]]: "one, two, three",
}
```

Admitedly, array literals are not likely to be ever used as keys in any conventional mapping type for obvious reasons. More likely is using an immutable list as a key:

```js
IMap#{
    [IList#[1, 2, 3]]: "one, two, three",
}
```

Disallowing barewords here means that we retain a certain symmetry - there is no possible syntax to use the bareword *to access* in general:

```js
const map = Map#{
    foo: "42",
};
map.get(foo);  // this is not valid
map.foo;  // neither is this
```

Finally, worth mentioning that Chrome (and presumably by extension Node) are currently dumping Maps to console like this:

```
> new Map([["1", "string one"], [1, "number one"]])
Map(2) {"1" => "string one", 1 => "number one"}
```

The choice of `=>` overloads the token for arrow functions, but it's also a possibility, similar to the alternative discussed in the proposal.

**Comment by alexweej** [comment]
Interestingly Firefox uses only colour to disambiguate string keys from number keys, and doesn't ever try to quote string keys!

```
> new Map([["1", "string one"], [1, "number one"], ["foo bar", "foo bar"]])
Map { 1: "string one", 1: "number one", foo bar: "foo bar" }
```

(Yeah, that's not good for copying and pasting, or people with colour deficiency.)

**Comment by bathos** [comment]
That alternative looks pretty sweet to me — the point that "there is no possible syntax to use the bareword to access in general" even gives it a certain elegant "oh yeah we meant to do that" feeling :)

**Comment by alexweej** [comment]
@bathos Which alternative? Requiring `[]` or using `=>`? (I assume from the remainder of your comment you mean `[]`, just checking...)

**Comment by ljharb** [comment]
Regular JS *is* "magic barewords with special exceptions" - that is how I expect it to work, and how I'd very strongly want it to work for consistency's sake. You also have an unproven premise here that this normal JS behavior is even bad.

Separately, there's no reason why `{ 1: 2 }` couldn't pass a `1` instead of a `'1'` to the Mappable object.

Using `->` or `=>` instead of `:` frees it from needing to be consistent, but simultaneously sacrifices precedence and the benefit of being an established pattern.

**Comment by alexweej** [comment]
@ljharb I am trying to use this opportunity to right a wrong for the continued health of the language - just try explaining the fact that "map literal keys are parsed differently to normal JS expressions" to any seasoned developer of another language and try not to be embarrassed as they laugh in your face! :)

If we insist on transplanting object literal's parsing quirks, then some might even argue that the existing array-of-two-element-arrays syntax *is better* than this proposal, as it does not invoke special parsing rules.

The case for object literals is much clearer, as I discuss in the proposal: `obj.foo` *is* the idiomatic way of accessing that property.

**Comment by alexweej** [comment]
> Separately, there's no reason why { 1: 2 } couldn't pass a 1 instead of a '1' to the Mappable object.

Interesting idea, but I think allowing *some* expressions through here just seems awkward because of the ambiguity around what `[1]` really means, as a key.

**Comment by bathos** [comment]
I meant the `[]` subset idea, yeah. It’s consistent, unambiguous, and still quite readable I think.

I’d agree that unquoted identifiers being strings in the context of map literals seems wacky. Unlike the case of property definition, there’s no special reason for this to be true; strings aren’t privileged in maps, so this tells the wrong story. I think there’s also some precedent for this sort of subsetting: object binding pattern syntax is a subset of object literal syntax, which in turn is almost-but-not-quite-because-of-commas a subset of class body syntax.

**Comment by GreenEyez86** [comment]
Agreed
